core/src/main/scala/kafka/controller/ZookeeperClient.scala

```diff
@@ -0,0 +1,287 @@
+    * @throws ZookeeperClientAuthFailedException if the authentication failed either before or while waiting for connection.
+    * @throws ZookeeperClientExpiredException if the session expired either before or while waiting for connection.
+    */
+  def waitUntilConnected(): Unit = inLock(isConnectedOrExpiredLock) {
+    waitUntilConnected(Long.MaxValue, TimeUnit.MILLISECONDS)
+  }
+
+  private def waitUntilConnected(timeout: Long, timeUnit: TimeUnit): Unit = {
+    info("Waiting until connected.")
+    var nanos = timeUnit.toNanos(timeout)
+    inLock(isConnectedOrExpiredLock) {
+      var state = zooKeeper.getState
+      while (!state.isConnected && state.isAlive) {
+        if (nanos <= 0) {
+          throw new ZookeeperClientTimeoutException(s"Timed out waiting for connection while in state: $state")
+        }
+        nanos = isConnectedOrExpiredCondition.awaitNanos(nanos)
+        state = zooKeeper.getState
+      }
+      if (state == States.AUTH_FAILED) {
+        throw new ZookeeperClientAuthFailedException("Auth failed either before or while waiting for connection")
+      } else if (state == States.CLOSED) {
+        throw new ZookeeperClientExpiredException("Session expired either before or while waiting for connection")
+      }
+    }
+    info("Connected.")
+  }
+
+  def registerZNodeChangeHandler(zNodeChangeHandler: ZNodeChangeHandler): ExistsResponse = {
+    registerZNodeChangeHandlers(Seq(zNodeChangeHandler)).head
+  }
+
+  def registerZNodeChangeHandlers(handlers: Seq[ZNodeChangeHandler]): Seq[ExistsResponse] = {
+    handlers.foreach(handler => zNodeChangeHandlers.put(handler.path, handler))
+    val asyncRequests = handlers.map(handler => ExistsRequest(handler.path, null))
+    handle(asyncRequests).asInstanceOf[Seq[ExistsResponse]]
+  }
+
+  def unregisterZNodeChangeHandler(path: String): Unit = {
+    zNodeChangeHandlers.remove(path)
+  }
+
+  def registerZNodeChildChangeHandler(zNodeChildChangeHandler: ZNodeChildChangeHandler): GetChildrenResponse = {
+    registerZNodeChildChangeHandlers(Seq(zNodeChildChangeHandler)).head
+  }
+
+  def registerZNodeChildChangeHandlers(handlers: Seq[ZNodeChildChangeHandler]): Seq[GetChildrenResponse] = {
+    handlers.foreach(handler => zNodeChildChangeHandlers.put(handler.path, handler))
+    val asyncRequests = handlers.map(handler => GetChildrenRequest(handler.path, null))
+    handle(asyncRequests).asInstanceOf[Seq[GetChildrenResponse]]
+  }
+
+  def unregisterZNodeChildChangeHandler(path: String): Unit = {
+    zNodeChildChangeHandlers.remove(path)
+  }
+
+  def close(): Unit = inWriteLock(initializationLock) {
+    info("Closing.")
+    zNodeChangeHandlers.clear()
+    zNodeChildChangeHandlers.clear()
+    zooKeeper.close()
+    info("Closed.")
+  }
+
+  private def initialize(): Unit = {
+    if (!zooKeeper.getState.isAlive) {
+      info(s"Initializing a new session to $connectString.")
+      var now = time.milliseconds()
```

Comment:
Yes, we need the timeout value. I am just not sure if we need a customized Time object. The main reason to have a Time is to mock it. Since it seems that we can't mock time in waitUntilConnected(), perhaps we don't really a Time object in the constructor of ZookeeperClient.